boa/src/builtins/object/internal_state.rs

```diff
@@ -45,20 +44,22 @@ impl Debug for InternalStateCell {
 
 impl InternalStateCell {
     /// Create new `InternalStateCell` from a value.
-    pub fn new<T: Any + InternalState>(value: T) -> Self {
+    pub fn new<T: 'static + InternalState>(value: T) -> Self {
         Self {
             state: Rc::new(value),
         }
     }
+
     /// Get a reference to the stored value and cast it to `T`.
     pub fn downcast_ref<T: Any + InternalState>(&self) -> Option<&T> {
         self.deref().downcast_ref::<T>()
     }
+
     /// Get a mutable reference to the stored value and cast it to `T`.
-    pub fn downcast_mut<T: Any + InternalState>(&mut self) -> Option<&mut T> {
-        self.deref_mut().downcast_mut::<T>()
+    pub fn downcast_mut<T: InternalState>(&mut self) -> Option<&mut T> {
+        self.state.downcast_mut::<T>()
     }
 }
 
 /// This trait must be implemented by all structs used for internal state.
-pub trait InternalState: Debug {}
+pub trait InternalState: Debug + Trace + Any {}
```

Comment:
It seems that this is not enough to be able to call `downcast_mut()` on `self.state`. It gives the following error:
```
method not found in `&(dyn builtins::object::internal_state::InternalState + 'static)`
```

Reply:
After some refactoring, I'm getting this error:
```
error[E0605]: non-primitive cast: `&dyn builtins::object::internal_state::InternalState` as `&(dyn std::any::Any + 'static)`
  --> boa/src/builtins/object/internal_state.rs:44:9
   |
44 |         (state as &dyn Any).downcast_ref::<T>()
   |         ^^^^^^^^^^^^^^^^^^^
   |
   = note: an `as` expression can only be used to convert between primitive types. Consider using the `From` trait

error[E0605]: non-primitive cast: `&mut (dyn builtins::object::internal_state::InternalState + 'static)` as `&mut (dyn std::any::Any + 'static)`
  --> boa/src/builtins/object/internal_state.rs:51:9
   |
51 |         (state as &mut dyn Any).downcast_mut::<T>()
   |         ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: an `as` expression can only be used to convert between primitive types. Consider using the `From` trait

```